x509 webflow doesn't handle authenticationFailure state

- CAS server version - 5.1 RC4-SNAPSHOT (built locally on about 4/21)
- Container/environment type/version - Tomcat, Windows
- To recreate: Authenticate with a validate certificate but principal lookup against attribute should fail
The following setting prevents principal from being used from certificate as a fallback: 
cas.authn.x509.principal.returnNull=true

I am working on a patch to handle the authenticationFailure state. When I tried sending flow to the login form (same flow as for ERROR) I get error due to this:

```
Caused by: org.springframework.expression.spel.SpelEvaluationException: EL1008E: Property or field 'username' cannot be found on object of type
 'org.apereo.cas.adaptors.x509.authentication.principal.X509CertificateCredential' - maybe not public?
```
Do I need create an action or do something to clear everything left over from X509 auth attempt out of the flow scope (or at least the credentials)?  Or should those credentials be removed when null is returned instead of the fallback to value from cert?

--- a/support/cas-server-support-x509-webflow/src/main/java/org/apereo/cas/web/flow/X509WebflowConfigurer.java
+++ b/support/cas-server-support-x509-webflow/src/main/java/org/apereo/cas/web/flow/X509WebflowConfigurer.java
@@ -8,7 +8,24 @@ import org.springframework.webflow.engine.builder.support.FlowBuilderServices;
 /**
  * The {@link X509WebflowConfigurer} is responsible for
  * adjusting the CAS webflow context for x509 integration.
- *
+ * 
+ * Creates a flow that starts by trying to construct credentials using an X509 
+ * certificate found as request attribute with key javax.servlet.request.X509Certificate
+ * {@link X509CertificateCredentialsNonInteractiveAction}
+ * If the check of the certificate is valid, flow goes to sendTicketGrantingTicket.
+ * On error or authenticationFailure, the user is sent to the login page. 
+ * The authenticationFailure outcome can happen when CAS got a valid certificate but 
+ * couldn't find entry for the certificate in an attribute repository and 
+ * falling back to principal from the certificate is turned off via:
+ * cas.authn.x509.principal.returnNull=true
+ * 
+ * Credentials are cleared out at the end of the action in case the user
+ * is sent to the login page where the X509 credentials object will cause
+ * errors (e.g. no username property) 
+ * 
+ * The X509 action is added to the main login flow by overriding the success
+ * outcome of the initializeLoginForm action.
+ * 
  * @author Misagh Moayyed
  * @since 4.2
  */
@@ -31,6 +48,9 @@ public class X509WebflowConfigurer extends AbstractCasWebflowConfigurer {
                     CasWebflowConstants.TRANSITION_ID_WARN));
             actionState.getTransitionSet().add(createTransition(CasWebflowConstants.TRANSITION_ID_ERROR,
                     CasWebflowConstants.STATE_ID_VIEW_LOGIN_FORM));
+            actionState.getTransitionSet().add(createTransition(CasWebflowConstants.TRANSITION_ID_AUTHENTICATION_FAILURE,
+                    CasWebflowConstants.STATE_ID_VIEW_LOGIN_FORM));
+
             actionState.getExitActionList().add(createEvaluateAction("clearWebflowCredentialsAction"));
             registerMultifactorProvidersStateTransitionsIntoWebflow(actionState);
 

